client/extensions/zoninator/components/search-autocomplete/suggestions.jsx

```diff
@@ -0,0 +1,167 @@
+
+		this.setState( {
+			suggestions,
+			suggestionPosition: 0,
+			currentSuggestion: first( suggestions ),
+		} );
+	}
+
+	filterSuggestions( suggestions, ignored ) {
+		if ( ! suggestions ) {
+			return [];
+		}
+
+		return suggestions.filter( ( { slug } ) => ! find( ignored, { slug } ) );
+	}
+
+	countSuggestions = () => this.state.suggestions.length;
+
+	getPositionForSuggestion = slug => findIndex( this.state.suggestions, { slug: slug } );
+
+	getSuggestionForPosition = position => this.state.suggestions[ position ];
+
+	increasePosition() {
+		const position = ( this.state.suggestionPosition + 1 ) % this.countSuggestions();
+		this.setState( {
+			suggestionPosition: position,
+			currentSuggestion: this.getSuggestionForPosition( position ),
+		} );
+	}
+
+	decreasePosition() {
+		const position = ( this.countSuggestions() + this.state.suggestionPosition - 1 ) % this.countSuggestions();
+		this.setState( {
+			suggestionPosition: position,
+			currentSuggestion: this.getSuggestionForPosition( position ),
+		} );
+	}
+
+	handleKeyEvent = ( event ) => {
+		if ( this.countSuggestions() === 0 ) {
+			return;
+		}
+
+		switch ( event.key ) {
+			case 'ArrowDown':
+				this.increasePosition();
+				event.preventDefault();
+				break;
+
+			case 'ArrowUp':
+				this.decreasePosition();
+				event.preventDefault();
+				break;
+
+			case 'Enter':
+				this.state.currentSuggestion && this.props.suggest( this.state.currentSuggestion );
+				break;
+		}
+	}
+
+	handleMouseDown = ( slug ) => {
+		const position = this.getPositionForSuggestion( slug );
+		this.props.suggest( this.getSuggestionForPosition( position ) );
+	}
+
+	handleMouseOver = ( slug ) => {
+		const position = this.getPositionForSuggestion( slug );
+		this.setState( {
+			suggestionPosition: position,
+			currentSuggestion: this.getSuggestionForPosition( position ),
+		} );
+	}
+
+	render() {
+		const {
+			searchTerm,
+			siteId,
+		} = this.props;
+
+		if ( ! searchTerm ) {
+			return null;
+		}
+
+		const showSuggestions = this.countSuggestions() > 0;
+		const suggestionsClass = 'zoninator__search-autocomplete__suggestions';
+
+		return (
+			<div>
+				<QueryPosts siteId={ siteId } query={ { search: searchTerm } } />
+
+				{
+					showSuggestions &&
+					<div className={ suggestionsClass }>
+						{ map( this.state.suggestions, ( post, idx ) => (
+							<SuggestionItem
+								key={ idx }
+								hasHighlight={ idx === this.state.suggestionPosition }
+								searchTerm={ searchTerm }
+								onMouseDown={ this.handleMouseDown }
+								onMouseOver={ this.handleMouseOver }
+								post={ post } />
+						) ) }
+					</div>
+				}
+			</div>
+		);
+	}
+}
+
+const mapStateToProps = ( state, { searchTerm } ) => {
+	const siteId = getSelectedSiteId( state );
+
+	return {
+		siteId,
+		suggestions: getSitePostsForQuery( state, siteId, { search: searchTerm } ),
+	};
+};
+
+const connectComponent = connect( mapStateToProps, null, null, { withRef: true } );
```

Comment:
Can we avoid the ref?

Reply:
I don't think we can, as it's used for passing down key events from the input and elevating all the logic to a parent component defeats the purpose in my opinion.